quotefault/__init__.py

```diff
@@ -215,28 +239,138 @@ def get():
 
     metadata = get_metadata()
-    metadata['submitter'] = request.args.get('submitter')  # get submitter from url query string
-    metadata['speaker'] = request.args.get('speaker')  # get speaker from url query string
-
-    #return quotes according to query strings (or lack thereof)
-    if metadata['speaker'] is not None and metadata['submitter'] is not None:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).filter(Quote.submitter == metadata['submitter'], Quote.speaker == metadata['speaker']).all()
-    elif metadata['submitter'] is not None:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).filter(Quote.submitter == metadata['submitter']).all()
-    elif metadata['speaker'] is not None:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).filter(Quote.speaker == metadata['speaker']).all()
-    else:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).all()
+
+    # Get all the quotes
+    quotes = get_quote_query(speaker = request.args.get('speaker'),
+        submitter = request.args.get('submitter')).all()
 
     #tie any votes the user has made to their uid
     user_votes = db.session.query(Vote).filter(Vote.voter == metadata['uid']).all()
 
-
     return render_template(
         'bootstrap/additional_quotes.html',
         quotes=quotes[20:],
         metadata=metadata,
         user_votes=user_votes
     )
+
+@app.route('/report/<report_id>', methods=['POST'])
+@auth.oidc_auth
+def submit_report(report_id):
+    """
+    Report a quote and notify EBoard/RTP
+    """
+    metadata = get_metadata()
+    existing_report = Report.query.filter(Report.reporter==metadata['uid'],
+        Report.quote_id==report_id).first()
+    if existing_report:
+        flash("You already submitted a report for this Quote!")
+        return redirect('/storage')
+    new_report = Report(report_id, metadata['uid'], None)
+    db.session.add(new_report)
+    db.session.commit()
+    if app.config['MAIL_SERVER'] != '':
+        send_report_email( metadata['uid'], Quote.query.get(report_id) )
+
+    flash("Report Successful!")
+    return redirect('/storage')
+
+@app.route('/review', methods=['GET'])
+@auth.oidc_auth
+def review():
+    """
+    Presents page for admins to review reports
+    and either hide or keep quote
+    """
+    metadata = get_metadata()
+    if metadata['is_admin']:
+        # Get the most recent 20 quotes
+        reports = Report.query.filter(Report.reviewed == False).all()
+
+        return render_template(
+            'bootstrap/admin.html',
+            reports=reports,
+            metadata=metadata
+        )
+    return redirect('/')

```

Comment:
Shouldn't this be a 403 or something?

Reply:
I would still want it to redirect instead of just crashing because it looks better, would this be a time to use `redirect('/',403)`?

Reply:
I mean, it doesn't have to crash, we can give a nice error page.

Reply:
I think for now, I can make it redirect to /, use a 403, and maybe flash a "Hey, you're not supposed to go there!" message

Reply:
Okay but that should be implemented as a flask error handler I think